### PR TITLE
lua+js: Implement and document utils.getpid()

### LINKED ISSUE #5208: mp.utils.getpid
This is a proposal for a new util function.

## Use case:

Scripts can launch external processes. Being able to tell that external process which mpv PID we're calling from (such as `os.execute("something " .. mp.utils.getpid())`) would allow that subprocess to differentiate the correct calling process if multiple mpv binaries are running.

In some cases, such as if we're calling an external C binary, we could of course just write that C binary to say "get the PID of the parent process". But in other cases, we don't have that luxury. For example, Apple's AppleScript has many great features for controlling app windows (via the Accessibility API) based on their process ID, but has no way of getting the calling process ID.

## Implementation:

- Bind to `getpid()` POSIX call on Mac and Linux (http://www.manpages.info/macosx/getpid.2.html).
- Alternative on Mac (but same result as `getpid()`): https://developer.apple.com/documentation/foundation/nsprocessinfo
- Bind to `GetCurrentProcessId()` on Windows: https://msdn.microsoft.com/en-us/library/windows/desktop/ms683180(v=vs.85).aspx

### PR DESCRIPTION
**Pull request by tsaaristo**
PIDs for the script kids.

This *has* been tested on Linux (Ubuntu) for both Lua and Javascript. Works, as should OSX.
This has *not* been tested on Windows, not yet. I'm working on getting builds on Windows happening, because this really needs to be tested on Windows - the script bindings are trivial, the real kicker is in replacing getpid() while on Windows. By what theory I've seen, it *should* work but I can't be certain, as I don't really do C.

Closes #5208.

### CODE REVIEW

**Comment by tsaaristo** [comment]
One confirmation from IRC that a cross-compiled mpv does work on Windows!

**Comment by avih** [comment]
No. There's no need for it. Next time please ask on irc or in a new issue if this is something we'd like to have. You do realize the barrier for most features is not writing the code itself but rather deciding whether or not it's a useful feature to have, right?

**Comment by tsaaristo** [comment]
@avih ?
But #5208 was already created for this feature, because it was discussed on the IRC. Why should I have made a duplicate issue for this? What's wrong with having the implementation ready?
I can see getting the current PID being rather useful with the more advanced scripts - this does not add bloat or risk or compromise some other feature. Who get to decide it's useless?
And are you really serious about the feature barrier being ideas rather than the execution? Ideas are plentiful, implementation requires effort.

**Comment by avih** [comment]
> But #5208 was already created for this feature, because it was discussed on the IRC. Why should I have made a duplicate issue for this?

I haven't noticed this issue, and you shouldn't have duplicated it, but rather added to it and ask if it's something we'd like to have.

A feature request is not automatically something we'd like to have. It's a feature _someone_ would like to have, and up to us to decide whether or not we want to add it.

> And are you really serious about the feature barrier being ideas rather than the execution? Ideas are plentiful, implementation requires effort.

I am. Good ideas are not plentiful, and it's true that coding requires an effort as well, but many times once a good idea is realized, then it's not hard to follow up with the technical details.

Either way, the point is that even if the code is good, it also needs to be something we want. And you didn't even bother to ask if we want it.

**Comment by tsaaristo** [comment]
I didn't know it was *so offensive to offer a PR*. What is this about?

I didn't "bother to ask" if it was desired because the issue is three days old, was tagged by sfan5 as an enhancement (which I gauged as passing some sort of rudimentary idea filter), and you had asked in IRC:
```
<avih> Yewbacca: TheAMM: who of you opened the mp.utils.getpid issue?
```
which made me believe you had had a look at it.

**Comment by avih** [comment]
> I didn't know it was so offensive to offer a PR. What is this about?

It's not offensive at all, on the contrary. But having a PR also doesn't mean we must accept it, right?

> and you had asked in IRC

I have, because I didn't know what it's good for and wanted more details on irc. You definitely can't quote me saying that it'd be a great thing to have.

**Comment by shinchiro** [comment]
@avih I believe you should state strong reason why this PR rejected, rather than 'no need' reason

**Comment by tsaaristo** [comment]
Having a PR doesn't mean you must accept it, but in your two messages, you put emphasis on the need for me to ask before making a PR (*"Next time please ask on irc or in a new issue if this is something we'd like to have."/"And you didn't even bother to ask if we want it."*), like I did something wrong here.
I am perfectly fine with a "Thanks, but we've decided we don't need this", but this was not one. You *chastice* me for trying to contribute. I'd understand if it was messy, or if I broke something or did something entirely wild, but this is just a `getpid()`.
I am not quoting you saying it'd be a great thing to have - I am quoting you to point out you had been aware of #5208, had not closed it and now chew me out for sending a PR for the issue.

**Comment by avih** [comment]
> but in your two messages, you put emphasis on the need for me to ask before making a PR

Only because you went on about it. My first sentence was "we don't need it", and I still stand by it, and I wanted to point out what could be a better process - first getting some agreement that it'd be a good thing to have.


> I am perfectly fine with a "Thanks, but we've decided we don't need this"

Apologies for anything else which might got implied by my replies.

Thanks, but we've decided we don't need this.

**Comment by SteveJobzniak** [comment]
Let's keep discussing this on IRC.

**Comment by avih** [comment]
> @avih I believe you should state strong reason why this PR rejected, rather than 'no need' reason

You're probably right that it wasn't enough details. The details are that it's not generally useful - at least I haven't seen any convincing arguments which suggests it is, and we try to maintain a codebase which balances usefulness without feature/code bloat.

While it's true that this PR is not a lot of code, it's also not generally useful, therefore even this little code is not worth having for a feature which no one would use. Except maybe one person.

**Comment by avih** [comment]
And for the record, I wouldn't mind opening this PR if an agreement emerges that this would be a useful thing to add.

**Comment by SteveJobzniak** [comment]
@avih It seems like there are 2 proposed uses for it so far:

- MacOS specific: Having the PID lets you tell AppleScript to do window management of mpv's player window, and the PID is necessary for finding the exact player instance when there are multiple ones.
- General (and this is the most interesting use case for me, now that I've heard about it): Having the PID lets you launch any external, detached process (`mp.utils.subprocess_detached({args:['somescript.py', mp.utils.getpid()]})`) and safely giving that script the ability to check if the launching parent PID is still alive. So even if mpv crashes or is force-killed, the subprocess, regardless of language, has a way of checking if the parent PID is still alive. Otherwise the subprocess would sit there in the background forever as an orphan.

There are surely even more uses for this.

**Comment by avih** [comment]
This PR is about code. Let's keep the discussion of whether it's required or not to #5208 .

**Comment by SteveJobzniak** [comment]
Ok, moving it there.
